lib/grpc_reflection/service/builder.ex

```diff
@@ -104,10 +121,78 @@ defmodule GrpcReflection.Service.Builder do
       root_symbols = %{symbol => response}
       root_files = %{(symbol <> ".proto") => response}
 
-      {%Agent{files: root_files, symbols: root_symbols}, referenced_types}
+      extension_file = symbol <> "Extension.proto"
+
+      {root_extensions, root_files} =
+        process_extensions(mod, symbol, extension_file, descriptor, root_files)
+
+      {%Agent{extensions: root_extensions, files: root_files, symbols: root_symbols},
+       referenced_types}
     end)
   end
 
+  # Processes extensions recursively for proto2, if any. Invoked only when extensions are present.
+  defp process_extensions(mod, symbol, extension_file, descriptor, root_files) do
+    with {:ok, {extension_numbers, extension_payload}} <-
+           process_extensions(mod, symbol, extension_file, descriptor) do
+      {
+        %{symbol => extension_numbers},
+        Map.put(root_files, extension_file, %{file_descriptor_proto: [extension_payload]})
+      }
+    else
+      {:ignore, _} -> {%{}, root_files}
+    end
+  end
+
+  defp process_extensions(
+         mod,
+         symbol,
+         extension_file,
+         %Google.Protobuf.DescriptorProto{extension_range: extension_range} = descriptor
+       )
+       when extension_range != [] do
+    unencoded_extension_payload = %Google.Protobuf.FileDescriptorProto{
+      name: extension_file,
+      package: package_from_name(symbol),
+      dependency: [symbol <> ".proto"],
+      syntax: get_syntax(mod)
+    }
+
+    {extension_numbers, extension_files} =
+      Enum.unzip(
+        for %Google.Protobuf.DescriptorProto.ExtensionRange{
+              start: start_index,
+              end: end_index
+            } <- descriptor.extension_range,
+            index <- start_index..end_index,
+            {_, ext} <- List.wrap(Protobuf.Extension.get_extension_props_by_tag(mod, index)) do
+          {index, Util.convert_to_field_descriptor(symbol, ext)}
+        end
+      )
+
+    message_list =
+      for ext <- extension_files, is_message_descriptor?(ext) do
+        ext.type_name
+        |> convert_symbol_to_module()
+        |> then(& &1.descriptor())
+      end
+
+    unencoded_extension_payload = %{
+      unencoded_extension_payload
+      | extension: extension_files,
+        message_type: message_list
+    }
+
+    {:ok, {extension_numbers, FileDescriptorProto.encode(unencoded_extension_payload)}}
+  end
+
+  defp process_extensions(_, _, _, _), do: {:ignore, {nil, nil}}
+
+  # according to Google.Protobuf.FieldDescriptorProto.Type, 11 is a message
+  defp is_message_descriptor?(%Google.Protobuf.FieldDescriptorProto{type: 11}), do: true
+
+  defp is_message_descriptor?(_), do: false
```

Comment:
Incidentally, we can access the enum values directly instead of hard-coding 11
`Google.Protobuf.FieldDescriptorProto.Type.mapping()[:TYPE_MESSAGE]`